loadbot: Allow use with all markets.

JoeGruffins opened the pull request:
closes #1565

chappjc commented on dex/testing/loadbot/loadbot.go:
```diff
@@ -1,6 +1,8 @@
 // This code is available on the terms of the project LICENSE.md file,
 // also available online at https://blueoakcouncil.org/license/1.0.0.
 
+//go:build lgpl
```
I'm fine with loadbot just using eth without the tag.  It's not a package that someone might import, nor an app we'd want to distribute, just a dev tool, so as devs we're fine with this.

JoeGruffins replied:
I think that registering the asset `_ "decred.org/dcrdex/client/asset/eth"` forces us to build with `--lgpl` anyhow.

chappjc commented:
Let's update loadbot's go.mod

```go
module decred.org/dcrdex/dex/testing/loadbot

go 1.17

// go mod tidy while commented:
// replace decred.org/dcrdex => ../../../

require (
	decred.org/dcrdex v0.0.0-20220609033337-f9b5b4ccdfac
	github.com/Shopify/toxiproxy/v2 v2.4.0
)
```

`go mod tidy -go=1.16 && go mod tidy -go=1.17`

Just need to update the toxiproxy import path:

```diff
diff --git a/dex/testing/loadbot/loadbot.go b/dex/testing/loadbot/loadbot.go
index 2032fd2b..7f65966c 100644
--- a/dex/testing/loadbot/loadbot.go
+++ b/dex/testing/loadbot/loadbot.go
@@ -32,7 +32,7 @@ import (
        "decred.org/dcrdex/dex/config"
        dexbtc "decred.org/dcrdex/dex/networks/btc"
        dexsrv "decred.org/dcrdex/server/dex"
-       toxiproxy "github.com/Shopify/toxiproxy/client"
+       toxiproxy "github.com/Shopify/toxiproxy/v2/client"
 )
```

JoeGruffins commented:
Working mostly for all coins. Some amounts still need to be adjusted, but I'm unsure of what's best. I think they can be adjusted as others see fit. For example, the heavy suite needs lots of funds that some harnesses do not have enough of.

The Zec compound tests fails with:
```
2022-06-15 21:03:38.573 [ERR] LOADBOT: exec error (zec) "./alpha": exit status 4: error code: -4
error message:
Error: The transaction was rejected! This might happen if some of the coins in your wallet were already spent, such as if you used a copy of the wallet and coins were spent in the copy but not marked as spent here.
```

All compound tests involving dcr seem to fail with this eventually: 
```
2022-06-15 20:40:41.371 [CRT] MANTLE:CMPD:STACKER:0: Error note received: {"type":"order","topic":"ReportRedeemError","subject":"Redeem reporting error","details":"Error notifying DEX of redemption for match 191411ecd922b23c003031a953fad008510c45290c4dfb381483c648a86178ae: error sending 'redeem' message: timed out waiting for \"redeem\" response","severity":5,"stamp":1655293241370,"acked":false,"id":"2325ea84f02ecc10","order":{"host":"127.0.0.1:17273","baseID":42,"baseSymbol":"dcr","quoteID":0,"quoteSymbol":"btc","market":"dcr_btc","type":1,"id":"3f84790cdc281dd8022081c36216580dac60314a5f4af1c94d3c55a5df28b073","stamp":1655293098911,"sig":"3045022100afc0a15cee1dc1619f3f7f432d52372586515e5148838ce4cfd8a0efc91c5c1402205f25ff00c360c016d39d718d50eba762ffa079101cb2d6152aae9428bdd2ce92","status":3,"epoch":110352873,"qty":9000000000,"sell":true,"filled":9000000000,"matches":[{"matchID":"191411ecd922b23c003031a953fad008510c45290c4dfb381483c648a86178ae","status":3,"active":false,"revoked":true,"rate":96600,"qty":6000000000,"side":0,"feeRate":10,"swap":{"id":"d7d55a9e238d0cfd05886fb040d57e0140d382a98cd89a6da65903b55bb9740a00000000","stringID":"0a74b95bb50359a66d9ad88ca982d340017ed540b06f8805fd0c8d239e5ad5d7:0","assetID":42,"symbol":"dcr"},"counterSwap":{"id":"76063548e54bee65f8e3678e0a58ff4ddf967db579b7f1388b985ec53bdabeb800000000","stringID":"b8beda3bc55e988b38f1b779b57d96df4dff580a8e67e3f865ee4be548350676:0","assetID":0,"symbol":"btc"},"redeem":{"id":"594496723fb9e309cde37781b45de9922ce2068d3ab3725913adec3a5f54882400000000","stringID":"2488545f3aecad135972b33a8d06e22c92e95db48177e3cd09e3b93f72964459:0","assetID":0,"symbol":"btc"},"stamp":1655293140000,"isCancel":false},{"matchID":"3c100d3ac61e20d03cbf4f1f5e7a22e437cba91e26116177f2490e3205ea16f8","status":3,"active":true,"revoked":false,"rate":96600,"qty":3000000000,"side":0,"feeRate":10,"swap":{"id":"9c4464fdb63b7ea13d4763f7967f962307a28cc23b500bd3848a83271c8c5db400000000","stringID":"b45d8c1c27838a84d30b503bc28ca20723967f96f763473da17e3bb6fd64449c:0","assetID":42,"symbol":"dcr"},"counterSwap":{"id":"c0a217da04ef01c00dc2d9c8b18342fd628f0571fffdbcc2dfeda59815506eb700000002","stringID":"b76e501598a5eddfc2bcfdff71058f62fd4283b1c8d9c20dc001ef04da17a2c0:2","assetID":0,"symbol":"btc"},"redeem":{"id":"8d3ccee98c9b412150dddec33ec82aa95ead5a36a3882676ebf7f61ef0945d4300000000","stringID":"435d94f01ef6f7eb762688a3365aad5ea92ac83ec3dedd5021419b8ce9ce3c8d:0","assetID":0,"symbol":"btc"},"stamp":1655293155000,"isCancel":false}],"cancelling":false,"canceled":false,"feesPaid":{"swap":4990,"redemption":286},"fundingCoins":[{"id":"11288f3c24c34fb2a1cb29660ba4634af38a12b36df44cfeada1e7b40374e3a900000001","stringID":"a9e37403b4e7a1adfe4cf46db3128af34a63a40b6629cba1b24fc3243c8f2811:1","assetID":42,"symbol":"dcr"}],"lockedamt":0,"accelerationCoins":[],"rate":96600,"tif":1,"targetOrderID":""}}
```

Other than that I think this pr solves the issue.

JoeGruffins commented on dex/testing/loadbot/mantle.go:
```diff
@@ -389,17 +466,20 @@ func coreLimitOrder(sell bool, qty, rate uint64) *core.TradeForm {
 // midGap parses the provided order book for the mid-gap price. If the book
 // is empty, a default value of rateStep * midGapMultiplier is returned instead.
 func midGap(book *core.OrderBook, rateStep uint64) uint64 {
+	if keepMidGap {
+		return uint64(defaultMidGap * 1e8)
+	}
 	if len(book.Sells) > 0 {
 		if len(book.Buys) > 0 {
-			return encodeRate((book.Buys[0].Rate + book.Sells[0].Rate) / 2)
+			return (book.Buys[0].MsgRate + book.Sells[0].MsgRate) / 2
```
This difference had me confused for a while. Cold not understand why the mid gap for eth would be so messed up.

chappjc commented on dex/testing/loadbot/compound.go:
```diff
@@ -36,18 +39,26 @@ func runCompound() {
-	var dcrPer uint64 = 15000e8 / 10 / 5
+	// Gotta get some major funding to the beta wallets.
+	var base uint64 = 100 * lotSize
+	var quote uint64 = calc.BaseToQuote(uint64(defaultMidGap*1e8), base)
```
There are a lot of `1e8` literals I assume are the `calc.RateEncodingFactor`, so should use that if that's the intent.

JoeGruffins replied:
Using a constant now.

chappjc commented on dex/testing/loadbot/mantle.go:
```diff
@@ -467,8 +547,60 @@ func newBotWallet(symbol, node, name string, pass []byte, minFunds, maxFunds uin
 	case btc:
 		form = &core.WalletForm{
-			Type:    "bitcoindRPC",
+			Type:    "SPV",
 			AssetID: btcID,
+			Config: map[string]string{
+				"rpcport": rpcAddr(symbol, node),
```
Two questions: Shouldn't we be able to test with bitcoindRPC?  In SPV, what's this `rpcport` setting for?

JoeGruffins replied:
> Shouldn't we be able to test with bitcoindRPC?

Yes. I just thought that most users will be using SPV so it's better to "test" with that. Will change to RPC tho.

> In SPV, what's this rpcport setting for?

Not needed...

chappjc replied:
> Will change to RPC tho

Well if I had to pick one to use on loadbot it would depend on what I'm hoping to test.  For client I agree SPV is best.  If aiming to just load the server, client might as well be super fast with full node.

Just thinking aloud

JoeGruffins replied:
Change to rpc.

chappjc commented:
> The Zec compound tests fails with:
> 
> ```
> 2022-06-15 21:03:38.573 [ERR] LOADBOT: exec error (zec) "./alpha": exit status 4: error code: -4
> error message:
> Error: The transaction was rejected! This might happen if some of the coins in your wallet were already spent, such as if you used a copy of the wallet and coins were spent in the copy but not marked as spent here.
> ```

Hmm, I guess that `./alpha` command it was trying to run was sending funds to the wallet to fund more orders?

Could this relate to restrictions in spending coinbase outputs?  https://github.com/zcash/zcash/pull/1431
https://github.com/zcash/zcash/issues/1373

JoeGruffins commented:
Just rebased.

JoeGruffins commented on dex/testing/loadbot/loadbot.go:
```diff
@@ -366,17 +536,20 @@ func main() {
+			case ethID:
+				oldPort := pair.cfg["ListenAddr"]
+				walletAddr = fmt.Sprintf("127.0.0.1%d", oldPort)
+				pair.cfg["ListenAddr"] = fmt.Sprintf(":%d", port)
```
I actually hadn't tested the latency switches. Will now...

JoeGruffins replied:
> port is a string but you have %d

fixed

> The walletAddr assignment looks off too.

In the conf file it is written like:
```
...
BootstrapNodesV5 = []
ListenAddr = ":${NODE_PORT}"
NetRestrict = [ "127.0.0.1/8", "::1/128" ]
...
```

It looks like it's working. There's no errors for me anyway, and toxiproxy is outputting stuff.

JoeGruffins commented on dex/testing/loadbot/mantle.go:
```diff
@@ -389,17 +466,20 @@ func coreLimitOrder(sell bool, qty, rate uint64) *core.TradeForm {
 // is empty, a default value of rateStep * midGapMultiplier is returned instead.
 func midGap(book *core.OrderBook, rateStep uint64) uint64 {
+	if keepMidGap {
+		return uint64(defaultMidGap * 1e8)
```
Using the constant now.

JoeGruffins commented on dex/testing/loadbot/loadbot.go:
```diff
@@ -23,120 +25,146 @@ import (
+	// Ports start from a random number so multiple loadbots may run at once.
+	// TODO: Find a better way to assign unused ports. This may fail randomly.
+	atomicPort                   = uint32(startPort + rand.Intn(10000))
 	conventionalConversionFactor = float64(dexbtc.UnitInfo.Conventional.ConversionFactor)
```
Removed.

JoeGruffins commented on dex/testing/loadbot/compound.go:
```diff
@@ -36,18 +39,26 @@ func runCompound() {
-			<-harnessCtl(dcr, "./alpha", "sendtoaddress", betaAddrDCR, valString(dcrPer))
+			if err := send(baseSymbol, alpha, betaAddrBase, quote); err != nil {
+				panic(fmt.Errorf("unable to send funds to quote asset: %v", err))
+				return
```
Removed.

JoeGruffins commented:
Diff since last review https://github.com/decred/dcrdex/compare/2b68ed650999f2654f8624b77a2309b874330d3e..67190d7e42b1b1ba647c28f3894f2db53eef13e9

JoeGruffins commented:
> Could this relate to restrictions in spending coinbase outputs?

It looks like the issue is fixed in that pr? Maybe just trying to send to many coins too fast? I don't see this with just the pingpong suite.

chappjc reviewed (approved):
Working well for me.

JoeGruffins commented on dex/testing/dcr/harness.sh:
```diff
@@ -219,12 +219,6 @@ tmux kill-session
 EOF
 chmod +x "${NODES_ROOT}/harness-ctl/quit"
 
-cat > "${NODES_ROOT}/harness-ctl/new-wallet" <<EOF
```
It was only one rpc call, so I didn't think it had a good purpose for the coins it existed in. The new start-wallet script has a lot of set-up involving ports already defined in the harness so I sorta moved this there for places where I thought it was needed.

chappjc replied:
Meh I think we can live without new-wallet. 
Descriptor wallets are going to be fine really soon anyway

JoeGruffins replied:
I added it back for btc, bch, and litecoin and think it's useful... Can remove again tho.

martonp reviewed (approved):
Just some very minor nits. I hadn't looked at the loadbot before.. we should add support for each of the different wallet types as well.

JoeGruffins commented:
Added back new-wallet for bitcoin assets and added info to the markets error. https://github.com/decred/dcrdex/compare/67190d7e42b1b1ba647c28f3894f2db53eef13e9..04faf33b6b728330d8afc15ee9bbf28402c2b73b

JoeGruffins commented on dex/testing/loadbot/mantle.go:
```diff
@@ -36,7 +38,7 @@ type Trader interface {
 // runTrader is the LoadBot workhorse. Creates a new mantle and runs the Trader.
 // runTrader will block until the ctx is canceled.
 func runTrader(t Trader, name string) {
-	m, err := NewMantle(name)
```
Removing `It is an error if a Decred wallet is not created by SetupWallets.`

JoeGruffins commented on dex/testing/loadbot/mantle.go:
```diff
@@ -192,7 +211,7 @@ type orderReq struct {
 // caller can target n orders per epoch with a dur = mkt.EpochDuration, and
 // avoid the simultaneous order at the beginning of the next epoch.
-// order places an order on the dcr_btc market.
+// order places an order on the market.
```
Removing.

JoeGruffins commented on dex/testing/loadbot/mantle.go:
```diff
@@ -231,7 +250,7 @@ func (m *Mantle) orderMetered(ords []*orderReq, dur time.Duration) {
 }
 
-// order places an order on the dcr_btc market.
+// order places an order on the market.
```
Changed.

JoeGruffins commented on dex/testing/loadbot/mantle.go:
```diff
@@ -300,12 +373,35 @@ func (m *Mantle) createWallet(symbol, node string, minFunds, maxFunds uint64, nu
+	case btc, dcr, ltc, doge, bch, zec:
+		res = <-harnessCtl(ctx, symbol, fmt.Sprintf("./%s", node), "sendtoaddress", addr, valString(val, symbol))
+	case eth:
+		// eth values are always handled as gwei, so multiply by nine
```
Fixed.

JoeGruffins commented:
Addressing marton's doc comments: https://github.com/decred/dcrdex/compare/04faf33b6b728330d8afc15ee9bbf28402c2b73b..493de785759483d6f9a481500af2c0ed26524d68

buck54321 reviewed (changes requested):
This is great. Just the one request.

JoeGruffins commented on dex/testing/loadbot/loadbot.go:
```diff
@@ -217,89 +307,170 @@ func main() {
 	if epochDuration == 0 {
-		fmt.Println("failed to find dcr_btc market in harness config")
-		return
+		return fmt.Errorf("failed to find %q market in harness config", market)
```
Now printing.

JoeGruffins commented:
Just rebased.

JoeGruffins commented:
Saw this for the first time on dcr_doge: `2022-06-20 14:13:19.230 [CRT] MANTLE:CMPD:STACKER:1: Trade error (limit order, sell = false, qty = 200000000, rate = 922000000): new order request with DEX server 127.0.0.1:17273 market dcr_doge failed: rpc error: error code 34: order failed validation`

JoeGruffins commented:
And:
```
2022-06-20 14:18:00.556 [CRT] MANTLE:CMPD:STACKER:1: Error note received: {"type":"order","topic":"SwapSendError","subject":"Swap send error","details":"Error encountered sending a swap output(s) worth 17.62000000 DOGE on order 8120c930","severity":5,"stamp":1655702280334,"acked":false,"id":"b435666548fbc052","order":{"host":"127.0.0.1:17273","baseID":42,"baseSymbol":"dcr","quoteID":3,"quoteSymbol":"doge","market":"dcr_doge","type":1,"id":"8120c9303f488661360b45c82534462cab40864979094e64e26adbb6d6330ff4","stamp":1655702239446,"sig":"3044022011979a725982a19ecbe4514ad8b7e21410c68a597fa53315216d799b30f45f0b02201c3aad4acd91bb7038fb124a9eda39ab49d3db87aa844e15334eff740b23f77f","status":3,"epoch":110380149,"qty":900000000,"sell":false,"filled":900000000,"matches":[{"matchID":"b14243ed2bcf822a19105e4a66d4677f9aa8ef81c7c8a9b07be0e35ecea41d2b","status":1,"active":true,"revoked":false,"rate":881000000,"qty":200000000,"side":0,"feeRate":4046,"swap":{"id":"926539219f312558a086d90243913405ba0013ad04a5bbfd67ab858764a9d09e00000000","stringID":"9ed0a9648785ab67fdbba504ad1300ba0534914302d986a05825319f21396592:0","assetID":3,"symbol":"doge","confs":{"count":1,"required":2}},"stamp":1655702265000,"isCancel":false},{"matchID":"08da6e681e54168732accbf3ebc32dee16c74e06142bfb0e6924293710f1576a","status":0,"active":true,"revoked":false,"rate":881000000,"qty":200000000,"side":0,"feeRate":4046,"stamp":1655702280000,"isCancel":false},{"matchID":"4c53b4f5b2653a56ad0bc329475b02f68e40ed7812607f5c2f92442e9601474e","status":1,"active":true,"revoked":false,"rate":881000000,"qty":400000000,"side":0,"feeRate":4046,"swap":{"id":"926539219f312558a086d90243913405ba0013ad04a5bbfd67ab858764a9d09e00000001","stringID":"9ed0a9648785ab67fdbba504ad1300ba0534914302d986a05825319f21396592:1","assetID":3,"symbol":"doge","confs":{"count":1,"required":2}},"stamp":1655702265000,"isCancel":false},{"matchID":"24e8c512cec8bdb93e0f61ffed2b4e512eed30f8ba6b72cd1be702c3efa6ef5e","status":1,"active":true,"revoked":false,"rate":881000000,"qty":100000000,"side":0,"feeRate":4046,"swap":{"id":"926539219f312558a086d90243913405ba0013ad04a5bbfd67ab858764a9d09e00000002","stringID":"9ed0a9648785ab67fdbba504ad1300ba0534914302d986a05825319f21396592:2","assetID":3,"symbol":"doge","confs":{"count":1,"required":2}},"stamp":1655702265000,"isCancel":false}],"cancelling":false,"canceled":false,"feesPaid":{"swap":1165248,"redemption":0},"fundingCoins":[{"id":"1d51ff124d747bccdd714ba0c09cc1b3a3e483ca7ad75ccf96eb23bb1bfb7d1300000000","stringID":"137dfb1bbb23eb96cf5cd77aca83e4a3b3c19cc0a04b71ddcc7b744d12ff511d:0","assetID":3,"symbol":"doge"}],"lockedamt":9717035824885984,"accelerationCoins":[],"rate":881000000,"tif":1,"targetOrderID":""}}
```

JoeGruffins commented:
And: `2022-06-20 14:38:46.457 [ERR] CORE:CMPD:STACKER:1: notify: |ERROR| (order) Swap reporting error - Error notifying DEX of swap for match 8403231ca1b8f5d2817238f04f69b3b4b307669b31f7214271ec09ecbbf362f7: error sending 'init' message: rpc error: error code 6: unknown match ID - Order: e9fb8f53af6538de3721e08f25456449b243d705e81978cb0500bd08d8703ce4
`

JoeGruffins commented:
Still seeing that last one even after building properly. Will make some issues I guess.

JoeGruffins commented:
Find unused ports: https://github.com/decred/dcrdex/compare/a68155e1aa3e76202b137dfa80604227016838cd..a2ad03d399015fc35e80a9ccc7fb66cc1de6292b

chappjc commented:
Regarding revoked matches in loadbot, while testing, I had wondered if it was even a good idea to use it with 30s/1m lock times rather than much higher.  I need to check how the new block mining is triggered from loadbot, and if it's possible matches will occasionally just be revoked if not settled fast enough.  (Given the recent proactive revoke changes to Swapper.)

chappjc commented:
No idea about the DOGE signing error.  Definitely make that an issue.

The order validation might just be a case of inadequate error detail - probably an order that was given a bad rate or quantity by loadbot.

JoeGruffins commented:
The mining is rather random. I guess they overlap too.

chappjc commented:
Nice!  I've run this PR a bunch and have little trouble aside from `compound` draining a wallet of funds after 90 min or so.

JoeGruffins commented:
Hmm, not sure how this can happen:
```
2022-06-20 15:53:00.348 [ERR] CORE:CMPD:STACKER:0[doge]: Withdraw error - address = 'n3iHHzBXe7QQTVKjCa7274DZsXMuDmJWfo', amount = -7572350.60830305: SendToAddress error: rawrequest (sendtoaddress) error: -3: Amount out of range
2022-06-20 15:53:00.349 [ERR] CORE:CMPD:STACKER:0: notify: |ERROR| (withdraw) Withdraw error - Error encountered during doge withdraw: SendToAddress error: rawrequest (sendtoaddress) error: -3: Amount out of range
2022-06-20 15:53:00.349 [CRT] MANTLE:CMPD:STACKER:0: failed to withdraw funds to alpha: SendToAddress error: rawrequest (sendtoaddress) error: -3: Amount out of range
```

chappjc commented:
> Can doge overflow

More likely underflow.  Kinda wish we were using signed int64 for amounts more broadly.
